firewood/src/proofs/diff.rs

```diff
@@ -0,0 +1,2028 @@
+/// needing to load the child from storage.
+struct TraversalStackFrame {
+    pre_path: Path,
+    node: Child,
+    hash: Option<TrieHash>,
+}
+
+/// State required for performing pre-order iteration of a Merkle trie. It contains
+/// a reference to a trie which implements the `TrieReader` trait, a stack that
+/// contains nodes (and their associated state including its pre-path and hash) to
+/// be traversed, and a previous number of children field.
+///
+/// The last field is used to specifically to pop off children from the last traversed
+/// node off of the stack so that they will not be traversed. This is useful in the
+/// case that the left trie node has the same path and hash as the right trie node. We
+/// can then skip their children in the traversal. `prev_num_children` is an u8 which
+/// is big enough since there can at most be 16 children per node.
+///
+/// Note that this skipping the children functionality can only be called once after a
+/// call to `next`. We cannot call `skip_children` twice in order to skip all of the
+/// children from the current node's parent. This functionality is not necessary for
+/// our use case as a hash match on the parent would have been determined previously
+/// since we are performing in-order traversal.
+struct PreOrderIterator<'a> {
+    trie: &'a dyn TrieReader,
+    stack: Vec<TraversalStackFrame>,
+    prev_num_children: u8,
+}
+
+/// Ignoring possible arithmetic side effects since we are only incrementing
+/// `prev_num_children` once per child and there can only be at most 16 children per
+/// node and `prev_num_children` is an u8.
+#[allow(clippy::arithmetic_side_effects)]
+impl PreOrderIterator<'_> {
+    // Creates a NodeState from a pre-path and Child that were popped off the
+    // traversal stack. The NodeState is used to compare nodes between tries.
+    fn create_node_state(
+        &self,
+        pre_path: &Path,
+        node: Child,
+        mut node_hash: Option<TrieHash>,
+    ) -> Result<NodeState, FileIoError> {
+        #[cfg(not(feature = "ethhash"))]
+        let node = match node {
+            Child::Node(node) => node.clone().into(),
+            Child::AddressWithHash(addr, hash) => {
+                node_hash = Some(hash);
+                self.trie.read_node(addr)?
+            }
+            Child::MaybePersisted(maybe_persisted, hash) => {
+                node_hash = Some(hash);
+                maybe_persisted.as_shared_node(&self.trie)?
+            }
+        };
+        #[cfg(feature = "ethhash")]
+        let node = match node {
+            Child::Node(node) => node.clone().into(),
+            Child::AddressWithHash(addr, hash) => {
+                node_hash = Some(hash.into());
+                self.trie.read_node(addr)?
+            }
+            Child::MaybePersisted(maybe_persisted, hash) => {
+                node_hash = Some(hash.into());
+                maybe_persisted.as_shared_node(&self.trie)?
+            }
+        };
```

Comment:
This code duplication can be avoided using the `into_triehash` method, which works correctly in both ethhash and non-ethhash modes.